Sources/Datadog/RUM/RUMVitals/VitalRefreshRateReader.swift

```diff
@@ -53,25 +55,53 @@ internal class VitalRefreshRateReader: ContinuousVitalReader {
         }
     }
 
+    // MARK: - Internal
+
+    internal func framesPerSecond(provider: FrameInfoProvider) -> Double? {
+        var fps: Double? = nil
+
+        if let lastFrameTimestamp = self.lastFrameTimestamp {
+            let currentFrameDuration = provider.currentFrameTimestamp - lastFrameTimestamp
+            guard currentFrameDuration > 0 else {
+                return nil
+            }
+            let currentFPS = 1.0 / currentFrameDuration
+
+            // ProMotion displays (e.g. iPad Pro and newer iPhone Pro) can have refresh rate higher than 60 FPS.
+            if let expectedCurrentFrameDuration = self.nextFrameDuration, provider.adaptiveFrameRateSupported {
+                guard expectedCurrentFrameDuration > 0 else {
+                    return nil
+                }
+                let expectedFPS = 1.0 / expectedCurrentFrameDuration
+                fps = currentFPS * (Self.backendSupportedFrameRate / expectedFPS)
+            } else {
+                fps = currentFPS
+            }
+        }
+
+        self.lastFrameTimestamp = provider.currentFrameTimestamp
+        self.nextFrameDuration = provider.nextFrameTimestamp - provider.currentFrameTimestamp
+
+        return fps
```

Comment:
/nit/suggestion

The nested statements make it difficult to catch, maybe we could flatten these by using a `defer`, e.g:
```swift
defer {
    self.lastFrameTimestamp = provider.currentFrameTimestamp
    self.nextFrameDuration = provider.nextFrameTimestamp - provider.currentFrameTimestamp
}

guard let lastFrameTimestamp = self.lastFrameTimestamp else {
    return nil
}

let currentFrameDuration = provider.currentFrameTimestamp - lastFrameTimestamp
guard currentFrameDuration > 0 else {
    return nil
}

let currentFPS = 1.0 / currentFrameDuration

// ProMotion displays (e.g. iPad Pro and newer iPhone Pro) can have refresh rate higher than 60 FPS.
guard provider.adaptiveFrameRateSupported, let expectedCurrentFrameDuration = self.nextFrameDuration else {
    return currentFPS
}

guard expectedCurrentFrameDuration > 0 else {
    return nil
}

let expectedFPS = 1.0 / expectedCurrentFrameDuration
return currentFPS * (Self.backendSupportedFrameRate / expectedFPS)
```

> If I got it right :)